controller/src/main/java/io/pravega/controller/store/index/ZKHostIndex.java

```diff
@@ -0,0 +1,218 @@
+import io.pravega.controller.store.stream.StoreException;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.api.BackgroundCallback;
+import org.apache.curator.framework.api.CuratorEvent;
+import org.apache.zookeeper.CreateMode;
+import org.apache.zookeeper.KeeperException;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.Random;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.stream.Collectors;
+
+/**
+ * Zookeeper based host index.
+ */
+@Slf4j
+public class ZKHostIndex implements HostIndex {
+    private final CuratorFramework client;
+    private final Executor executor;
+    private final String hostRoot;
+
+    public ZKHostIndex(CuratorFramework client, String hostRoot, Executor executor) {
+        this.client = client;
+        this.executor = executor;
+        this.hostRoot = hostRoot;
+    }
+
+    @Override
+    public CompletableFuture<Void> addEntity(final String hostId, final String entity) {
+        return addEntity(hostId, entity, new byte[0]);
+    }
+
+    @Override
+    public CompletableFuture<Void> addEntity(String hostId, String entity, byte[] entityData) {
+        Preconditions.checkNotNull(hostId);
+        Preconditions.checkNotNull(entity);
+        return createNode(CreateMode.PERSISTENT, true, getHostPath(hostId, entity), entityData);
+    }
+
+    @Override
+    public CompletableFuture<byte[]> getEntityData(String hostId, String entity) {
+        Preconditions.checkNotNull(hostId);
+        Preconditions.checkNotNull(entity);
+        return readNode(getHostPath(hostId, entity));
+    }
+
+    @Override
+    public CompletableFuture<Void> removeEntity(final String hostId, final String entity, final boolean deleteEmptyHost) {
+        Preconditions.checkNotNull(hostId);
+        Preconditions.checkNotNull(entity);
+        return deleteNode(getHostPath(hostId, entity)).thenCompose(ignore -> {
+            if (deleteEmptyHost) {
+                return deleteNode(getHostPath(hostId)).exceptionally(ex -> {
+                    if (ex instanceof StoreException.NodeNotEmptyException) {
+                        return null;
+                    } else {
+                        throw (StoreException) ex;
+                    }
+                });
+            } else {
+                return CompletableFuture.completedFuture(null);
+            }
+        });
+    }
+
+    @Override
+    public CompletableFuture<Void> removeHost(final String hostId) {
+        Preconditions.checkNotNull(hostId);
+        return deleteNode(getHostPath(hostId));
+    }
+
+    @Override
+    public CompletableFuture<Optional<String>> getRandomEntity(final String hostId) {
+        Preconditions.checkNotNull(hostId);
+        return getChildren(getHostPath(hostId)).thenApply(children -> {
```

Comment:
If all the implementation is going to do is get all of the Entities and return one at random then it might as well be replaced with a function that just returns all the entities and let the caller pick a random one. This is more general and preserves the abstraction. If there was a non-trivial performance benefit it might be worth the tradeoff but there isn't if it has to fetch them all anyway. If the concern is duplication of code, we could just create a utility function in Common that grabs a random element from a set.